src/Chapter4.hs

```diff
@@ -508,15 +508,17 @@ instance Applicative List where
   pure :: a -> List a
   pure x = Cons x Empty
 
+  (<*>) :: List (a -> b) -> List a -> List b
   (<*>) Empty _ = Empty
-  (<*>) fs ms =
-    goM Empty ms
+  (<*>) fs ms = loopM Empty ms fs
     where
-      goF acc Empty _ = acc
-      goF acc (Cons f fTail) m = goF (Cons (f m) acc) fTail m
+      loopM :: List b -> List a -> List (a -> b) -> List b
+      loopM acc Empty _ = acc
+      loopM acc (Cons mHead mTail) f = loopF (loopM acc mTail f) f mHead
 
-      goM acc Empty = acc
-      goM acc (Cons m mTail) = goF (goM acc mTail) fs m
+      loopF :: List b -> List (a -> b) -> a -> List b
+      loopF acc Empty _ = acc
+      loopF acc (Cons f fTail) m = Cons (f m) $ loopF acc fTail m
```

Comment:
Now it's a bit cleaner what's going on 🙂 So, my suggestion to simplify the implementation is the following:

1. In the `<*>` operator, handle two cases with `Empty` separately: `_ <*> Empty` and `Empty <*> _`.
2. After that, you will have a third case where the first pattern is `Cons`: `Cons f fs <*> xs`.
3. In that third case, you can apply the `f` function to `xs` and append to the result of `fs <*> xs`. For this, you need to implement a function for appending two lists: `append :: List a -> List a -> List a`.

I believe, this new final implementation will be cleaner and simpler 🙂

Reply:
hmm, let me try again